Fix classical bit mapping in HLS pass

ElePT opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary
This PR fixes a bug in HLS where classical registers were mapped to the indices in within the high level object they were defined on, instead of the corresponding index in the global circuit. 

Fixes #14569 . 


### Details and comments
The bug was introduced in 2.0, so the fix should be backported to the `stable/2.0` branch, but not `stable/1.4`.

ElePT committed:
Map classical bits from block to outer circuit

Co-authored-by: Alexander Ivrii <alexi@il.ibm.com>

ElePT committed:
Add regression test

ElePT committed:
Add reno

mtreinish commented on crates/transpiler/src/passes/high_level_synthesis.rs:
```diff
@@ -479,6 +479,11 @@ fn run_on_circuitdata(
+            .get_cargs(inst.clbits)
+            .iter()
+            .map(|c| c.index())
+            .collect::<Vec<usize>>();
```
This is already a `Clbit` why do you need to convert it to a usize and then back to a `Clbit` below. Couldn't you just do
```suggestion
            .get_cargs(inst.clbits)
            .to_vec();
```

and then change the `map()` below to `op_clbits[c.0 as usize]`.

mtreinish replied:
I just implemented this in: https://github.com/Qiskit/qiskit/pull/14597/commits/000253d633703708ff60edf625385f028d41e2f3 when I looked at the code again I also realized you don't need to store this in a vec at all since the slice will do. Looking at the qubits path the same optimization can mostly be applied too but since this is a fix for backport I didn't make that change here.

mtreinish committed:
Avoid allocating cargs

Previously the cargs were being written to a Vec<usize> as an
intermediate storage, however we are just returning them directly and
don't need to allocate or change the types. This commit just works with
the slice returned from the interner directly which both simplifies the
code and should also be faster.

mtreinish commented:
@mergifyio backport stable/2.0
